kpt-functions/replacement/function.go

```diff
@@ -192,29 +167,57 @@ func prepareValueFromMultiRef(items []*yaml.RNode, m *MultiSourceObjRef) (string
 	}
 
 	var out bytes.Buffer
-	tmpl, err := template.New("tmpl").Funcs(sprig.TxtFuncMap()).Parse(m.Template)
+	tmpl, err := template.New("tmpl").Funcs(sprig.TxtFuncMap()).Parse(mObjRef.Template)
 	if err != nil {
-		return "", fmt.Errorf("error parsing template %s: %w", m.Template, err)
+		return "", fmt.Errorf("error parsing template %s: %w", mObjRef.Template, err)
 	}
 
-	err = tmpl.Execute(&out, data)
-	if err != nil {
+	if err := tmpl.Execute(&out, data); err != nil {
 		return "", fmt.Errorf("error executing template %s, data %v: %w",
-			m.Template, data, err)
+			mObjRef.Template, data, err)
 	}
 
 	return out.String(), nil
 }
 
+func (s Source) prepareValue(items []*yaml.RNode) (interface{}, error) {
+	if s.Value != "" {
+		return s.Value, nil
+	}
+	if s.ObjRef != nil {
+		return s.ObjRef.prepareValue(items, s.FieldRef)
+	}
+	if s.MultiRef != nil {
+		return s.MultiRef.prepareValue(items)
+	}
+	// @aodinokov, Should it be an error?
+	return "", nil
+}
+
+func (f Function) Exec(items []*yaml.RNode) error {
```

Comment:
BTW,

in the last incarnations of plugins I started to call this as Filter function with a bit different signature:
https://github.com/aodinokov/kpt-functions-catalog/blob/function-with-env/functions/go/templater/function/filter.go#L42

this is to match kyaml interface Filter. in that case it's also possilble to use these functions from code inside kyaml filters.

probably something to think about